File: IPWorks Samples/POP Email Client/c++/popclient.cpp
/*
 * IPWorks 2024 C++ Edition - Sample Project
 *
 * This sample project demonstrates the usage of IPWorks in a 
 * simple, straightforward way. It is not intended to be a complete 
 * application. Error handling and other checks are simplified for clarity.
 *
 * www.nsoftware.com/ipworks
 *
 * This code is subject to the terms and conditions specified in the 
 * corresponding product license agreement which outlines the authorized 
 * usage and restrictions.
 */


#include <stdio.h>
#include <string.h>
#include <stdlib.h>
#include "../../include/ipworks.h"

#include <ctype.h>

#define LINE_LEN 80

enum { HEADERS, FULL, REPLY };

class MyPOP: public POP
{
public:
	MyPOP()
	{
		state = HEADERS;
		lines = 0;
		strcpy(reply,"");
	}

	virtual int FireHeader(POPHeaderEventParams *e)
	{

		if (state == FULL)
		{
			printf("%s: %s\n", e->Field, e->Value);
		}

		else if ( !strcmp(e->Field, "From") )
		{
			sprintf(msg_from, e->Value);
		}

		else if ( !strcmp(e->Field, "Date") )
		{
			sprintf(msg_date, e->Value);
		}

		else if ( !strcmp(e->Field, "Subject") )
		{
			sprintf(msg_subject, e->Value);
			if ( state == REPLY )      //  change the subject to Re:
			{
				char temp[LINE_LEN];
				sprintf(temp, msg_subject);
				strcpy(msg_subject, "Re: ");
				strcat(msg_subject, temp);
			}
		}
		return 0;
	}

	virtual int FireTransfer(POPTransferEventParams *e)
	{
		if (state == FULL)
		{
			printf("%s\n", e->Text);
			lines++;
			if ( lines == 22 )
			{
				printf("Press ENTER to continue...");
				getchar();
				lines = 0;
			}
		}
		else if (state == REPLY)     // if replying, add original message text to reply text
		{
			strcat(reply,e->Text);
			strcat(reply,"\n");
		}
		return 0;
	}

	virtual int FireSSLServerAuthentication(POPSSLServerAuthenticationEventParams *e)
	{
		if (e->Accept) return 0;
		printf("Server provided the following certificate:\nIssuer: %s\nSubject: %s\n",
			e->CertIssuer, e->CertSubject);
		printf("The following problems have been determined for this certificate: %s\n", e->Status);
		printf("Would you like to continue anyways? [y/n] ");
		if (getchar() == 'y') e->Accept = true;
		else exit(0);
		return 0;
	}

	int state;
	char reply[10000];
	char msg_from[LINE_LEN];
	char msg_date[LINE_LEN];
	char msg_subject[LINE_LEN];
	int lines;

	void clearHeadlineData()
	{
		msg_from[0] = '\0';
		msg_date[0] = '\0';
		msg_subject[0] = '\0';
	}

	void printHeadline()
	{
		printf("%-4d%-21.19s%-21.19s%-21.19s\n", GetMessageNumber(), msg_from, msg_date, msg_subject);
	}
};

class MyFileMailer: public FileMailer
{

public:
	MyFileMailer() { }

	virtual int FireTransfer(FileMailerTransferEventParams *e)
	{
		//printf("Transfer: %ld", e->BytesTransferred);
		return 0;
	}

};

int main(int argc, char * argv[])
{

	if (argc != 6) {
		fprintf(stderr, "* This demo shows how to use POP component to view messages in the mailbox and reply to an email with the Filemailer component. *\n");
		fprintf(stderr, "usage: popclient popserver user password smtpserver from\n");
		fprintf(stderr, "\n");
		fprintf(stderr, "  popserver   the name or address of a mail server (internet post office server)\n");
		fprintf(stderr, "  user        the user identifier for the mailbox\n");
		fprintf(stderr, "  password    the password for the mailbox user\n");
		fprintf(stderr, "  smtpserver  the name or address of a mail server (mail relay)\n");
		fprintf(stderr, "  from        the email address of the sender\n");
		fprintf(stderr, "\nExample: popclient mail.local username password mail.local user@service.com\n\n");
		printf("Press enter to continue.");
		getchar();

	}
	else{
		MyPOP pop;                  // POP object
		MyFileMailer smtp;                  // SMTP object
		int ret_code = 0;           // Return code:
		// = 0 when there is no error
		// = error code when there is an error

		char command[LINE_LEN];     // user's command
		char buffer[LINE_LEN];      // text buffer
		char *argument;            // arguments to the user's command

		int k = 0;         // counter for use in for() loops
		char msg[10000];            // storage for mail message
		int msgnum = 0;             // current message number for next command
		
		if (ret_code = pop.SetMailServer(argv[1])) goto done;		
		pop.SetUser(argv[2]);		
		pop.SetPassword(argv[3]);
		if (ret_code = pop.Connect()) goto done;		
		smtp.SetMailServer(argv[4]);		
		smtp.SetFrom(argv[5]);

		printf("Type \"?\" for a list of commands.\n\n");

		while (1)
		{

			strcpy(msg, "");
			printf("> ");
			fgets(command, LINE_LEN, stdin);
			command[strlen(command) - 1] = '\0';
			argument = strtok(command, " \t\n");

			if (!strcmp(command, "?"))
			{

				printf("Mail Commands\n"
					"t <message number>              type messages\n"
					"n                               goto and type next message\n"
					"f                               give head lines of messages\n"
					"d <message number>              delete messages\n"
					"m <user>                        mail to specific users\n"
					"q                               quit, saving unresolved messages in mbox\n"
					"h                               print out active message headers\n"
					"r <message number>              reply to message\n");

			}

			else if (!strcmp(command, "d"))
			{

				if (argument = strtok(NULL, " \t\n"))
				{
					msgnum = atoi(argument);
					pop.SetMessageNumber(msgnum);
					ret_code = pop.Delete();
				}

				else
				{
					printf("No message specified.\n");
				}

			}

			else if (!strcmp(command, "f"))
			{

				pop.state = HEADERS;
				pop.SetMaxLines(1);
				for (k = 1; k <= pop.GetMessageCount(); k++)     // get header info for every message
				{
					pop.SetMessageNumber(k);
					pop.clearHeadlineData();
					ret_code = pop.Retrieve();
					pop.printHeadline();
				}
				pop.SetMaxLines(0);

			}

			else if (!strcmp(command, "h"))     // get header info for every message
			{

				pop.state = HEADERS;
				pop.SetMaxLines(1);
				for (k = 1; k <= pop.GetMessageCount(); k++)
				{
					pop.SetMessageNumber(k);
					pop.clearHeadlineData();
					ret_code = pop.Retrieve();
					pop.printHeadline();
				}
				pop.SetMaxLines(0);

			}

			else if (!strcmp(command, "m"))
			{


				if (!(argument = strtok(NULL, " \t\n")))    // if no recipient specified
				{
					printf("To: ");
					fgets(buffer, LINE_LEN, stdin);
					buffer[strlen(buffer) - 1] = '\0';
					smtp.SetSendTo(buffer);
				}
				else
				{
					smtp.SetSendTo(argument);
				}

				printf("Subject: ");
				fgets(buffer, LINE_LEN, stdin);
				buffer[strlen(buffer) - 1] = '\0';
				smtp.SetSubject(buffer);

				printf("Type \".\" on a line by itself to end message.\n");
				strcpy(msg, "\n");                          // initialize the message string
				fgets(buffer, LINE_LEN, stdin);
				buffer[strlen(buffer) - 1] = '\0';                             // get first line
				while (strcmp(buffer, "."))                // until period on a line by itself
				{
					strcat(msg, buffer);                    // add a line
					strcat(msg, "\n");                       // with an endline
					fgets(buffer, LINE_LEN, stdin);
					buffer[strlen(buffer) - 1] = '\0';                         // get next line
				}

				if (ret_code = smtp.SetMessageText(msg)) goto done;

				printf("Cc: ");
				fgets(buffer, LINE_LEN, stdin);
				buffer[strlen(buffer) - 1] = '\0';
				smtp.SetCc(buffer);

				ret_code = smtp.Send();

			}

			else if (!strcmp(command, "n"))
			{

				msgnum++;
				pop.state = FULL;
				pop.SetMessageNumber(msgnum);
				ret_code = pop.Retrieve();

			}

			else if (!strcmp(command, "q"))
			{

				printf("Save changes to inbox? (y/n): ");
				if (getchar() == 'n')
				{
					ret_code = pop.Reset();
				}
				ret_code = pop.Disconnect();
				exit(0);

			}

			else if (!strcmp(command, "r"))
			{

				if (argument = strtok(NULL, " \t\n"))    // get number of message to reply to
				{
					msgnum = atoi(argument);
					printf("Type your message here. Type \".\" on a line by itself to continue.\n");
				}
				if (!msgnum)
				{
					printf("No message specified.\n");
				}
				else
				{
					pop.state = REPLY;
					pop.SetMessageNumber(msgnum);
					ret_code = pop.Retrieve();
					smtp.SetSendTo(pop.msg_from);
					smtp.SetSubject(pop.msg_subject);

					fgets(buffer, LINE_LEN, stdin);
					buffer[strlen(buffer) - 1] = '\0';
					while (strcmp(buffer, "."))
					{
						strcpy(msg, buffer);
						strcat(msg, "\n");
						fgets(buffer, LINE_LEN, stdin);
						buffer[strlen(buffer) - 1] = '\0';
					}

					strcat(msg, pop.msg_from);  // insert name of previous message author
					strcat(msg, " wrote: \n");
					strcat(msg, pop.reply);     // insert previous message
					strcpy(pop.reply, "");      // reset reply field for future use

					smtp.SetMessageText(msg);
					printf("Cc: ");
					fgets(buffer, LINE_LEN, stdin);
					buffer[strlen(buffer) - 1] = '\0';
					smtp.SetCc(buffer);
					ret_code = smtp.Send();
				}
			}

			else if (!strcmp(command, "t"))
			{

				if (argument = strtok(NULL, " \t\n"))
				{
					msgnum = atoi(argument);
					pop.state = FULL;
					pop.SetMessageNumber(msgnum);
					ret_code = pop.Retrieve();
				}

				else
				{
					printf("No message specified.\n");
				}

			}
			else if (!strcmp(command, ""))
			{
				// do nothing
			}
			else
			{
				if (isdigit((int)command[0]))
				{
					// allow user to enter only the number
					//  of the message they want to view
					msgnum = atoi(command);
					pop.state = FULL;
					pop.SetMessageNumber(msgnum);
					ret_code = pop.Retrieve();
				}
				else
				{
					printf("Bad command / Not implemented in demo.\n");
				}

			} // end of command checking

			if (ret_code)     // Got an error.  The user is done.
			{
				printf("\nError: %d", ret_code);
				if (pop.GetLastError())
				{
					printf(" \"%s\"\n", pop.GetLastError());
				}
				if (smtp.GetLastError())
				{
					printf(" \"%s\"\n", smtp.GetLastError());
				}
			}

			ret_code = 0;  // flush out the error

		}  // end of main while loop

	done:
		if (ret_code)     // Got an error.  The user is done.
		{
			printf("\nError: %d", ret_code);
			if (pop.GetLastError())
			{
				printf(" \"%s\"\n", pop.GetLastError());
			}
			if (smtp.GetLastError())
			{
				printf(" \"%s\"\n", smtp.GetLastError());
			}
			printf("Exiting... (press enter)\n");
			getchar();
			exit(1);
		}
		return 0;
	}
	
}

File: IPWorks Samples/POP Email Client/java/popclient.java
/*
 * IPWorks 2024 Java Edition - Sample Project
 *
 * This sample project demonstrates the usage of IPWorks in a 
 * simple, straightforward way. It is not intended to be a complete 
 * application. Error handling and other checks are simplified for clarity.
 *
 * www.nsoftware.com/ipworks
 *
 * This code is subject to the terms and conditions specified in the 
 * corresponding product license agreement which outlines the authorized 
 * usage and restrictions.
 */

import java.io.*;
import ipworks.*;

public class popclient extends ConsoleDemo {
    
	public static final int HEADERS = 1;
    public static final int FULL = 2;
    public static final int REPLY = 3;
    static int state;
    static String reply;
    static String msg_from;
    static String msg_date;
    static String msg_subject;
    static int msg_field;
    static int lines;
    static POP pop;
 
    public static void main(String[] args) {
    	
    	if(args.length!=5) {
			System.out.println("* This demo shows how to use POP component to view messages in the mailbox and reply to an email with the Filemailer component. *");
			System.out.println("usage: popclient popserver user password smtpserver from");
			System.out.println("");
			System.out.println("  popserver   the name or address of a mail server (internet post office server)");
			System.out.println("  user        the user identifier for the mailbox");
			System.out.println("  password    the password for the mailbox user");
			System.out.println("  smtpserver  the name or address of a mail server (mail relay)");
			System.out.println("  from        the email address of the sender ");
			System.out.println("\r\nExample: popclient mail.local username password mail.local user@service.com");
			
		} else {
			state = HEADERS;
			msg_field = 0;
			lines = 0;
			reply = "";
        
			pop = new POP();                // POP object
			SMTP smtp = new SMTP();   // SMTP object

			String command;     			// user's command
			String buffer;      			// text buffer
			String[] argument;        		// arguments to the user's command

			int k;         					// counter for use in for() loops
			String msg;            			// storage for mail message
			int msgnum = 0;             	// current message number for next command

			try {
				pop.addPOPEventListener(new DefaultPOPEventListener(){      		
					public void SSLServerAuthentication(POPSSLServerAuthenticationEvent e) {
						e.accept=true; //this will trust all certificates and it is not recommended for production use	
					}
					public void header(POPHeaderEvent e) {
						if (state == FULL){
							System.out.println(e.field+": "+e.value);
						} else if (e.field.equals("From") ){
							msg_from = e.value;
							msg_field++;
						} else if (e.field.equals("Date") ){       	        
							msg_date = e.value;
							msg_field++;
						}else if (e.field.equals("Subject") ){
							msg_subject = e.value;
							msg_field++;
							if ( state == REPLY ) {    //  change the subject to Re:
								String temp;
								temp = msg_subject;
								msg_subject = "Re: ";
								msg_subject += temp;
							}
						}
						if (state == HEADERS && msg_field == 3){
							System.out.print(pop.getMessageNumber()+" ");
							System.out.print(msg_from+" ");
							System.out.println(msg_date+" "+msg_subject);
							msg_field = 0;
						}
					}       		
					public void transfer(POPTransferEvent e) {
						String s = new String(e.text);
						if (state == FULL){
							System.out.println(s);
							lines++;
							if ( lines == 22 ){
								System.out.print("Press ENTER to continue...");
								try { System.in.read(); } catch(IOException ex) {}
								lines = 0;
							}
						} else if (state == REPLY) {   // if replying, add original message text to reply text
							reply += s;
							reply += "\n";
						}
					}        		      									
				});
            
	            pop.setMailServer( args[0] ); // pop server       
	            pop.setUser( args[1] ); // user
	            pop.setPassword( args[2] ); // password
	            pop.connect();
	
	            smtp.setMailServer( args[3]); // smtp server
	            smtp.setFrom( args[4] );
	
	            DisplayMenu();
        
	            while(true)
	            {
	                msg = "";
	                System.out.print( "> " );
	                command = input();
	                argument = command.split("\\s");
	
	                if ( argument.length == 0 || argument[0].length() == 0)
	                {
	                    // do nothing
	                }               
	                else if ( argument[0].equals("d") )
	                {
	                    if ( argument.length > 1 )
	                    {
	                        msgnum = Integer.parseInt(argument[1]);
	                        pop.setMessageNumber(msgnum);
	                        pop.delete();
	                    }
	                    else
	                    {
	                        System.out.println( "No message specified.");
	                    }
	                }
	                else if ( argument[0].equals("h") )
	                {   // get header info for every message
	                    state = HEADERS;
	                    pop.setMaxLines(1);
	                    for (k = 1; k <= pop.getMessageCount(); k++) {
	                        pop.setMessageNumber(k);
	                        pop.retrieve();
	                    }
	                    pop.setMaxLines(0);
	                }
	                else if ( argument[0].equals("m") )
	                {
	                    if ( argument.length < 2 )
	                    {  // if no recipient specified
	                        System.out.print( "To: " );
	                        buffer = input();
	                        smtp.setSendTo( buffer );
	                    }
	                    else
	                    {
	                        smtp.setSendTo(argument[1]);
	                    }
	
	                    System.out.print("Subject: ");
	                    buffer = input();
	                    smtp.setSubject( buffer );
	                                   
	                    System.out.println("Type your message here.Type \".\" on a line by itself to end message.");
	                    msg = "\n";                          // initialize the message string
	                    buffer = input();                    // get first line
	                    while( !buffer.equals(".") )
	                    {              // until period on a line by itself
	                        msg += buffer;                    // add a line
	                        msg += "\n";                      // with an endline
	                        buffer = input();                 // get next line
	                    }
	
	                    smtp.setMessageText(msg);
	
	                    System.out.print( "Cc: " );
	                    buffer = input();
	                    smtp.setCc(buffer);
	
	                    smtp.send();
	                }
	                else if ( argument[0].equals("n") )
	                {
	                    msgnum++;
	                    state = FULL;
	                    pop.setMessageNumber(msgnum);
	                    pop.retrieve();
	                }
	                else if ( argument[0].equals("q") )
	                {
	                    System.out.print("Save changes to inbox? (y/n): ");
	                    int n = System.in.read();
	                    if ( n == 'n' )
	                    {
	                        pop.reset();
	                    }
	                    pop.disconnect();
	                    System.exit(0);
	                    return;
	                }
	                else if ( argument[0].equals("r") )
	                {
	                    if ( argument.length > 1 )
	                    {  // get number of message to reply to
	                        msgnum = Integer.parseInt(argument[1]);
	                        System.out.println("Type your message here. Type \".\" on a line by itself to continue.");
	                    }
	                    if(msgnum == 0)
	                    {
	                        System.out.println("No message specified.");
	                    }
	                    else
	                    {
	                        state = REPLY;
	                        pop.setMessageNumber(msgnum);
	                        pop.retrieve();
	                        smtp.setSendTo(msg_from);
	                        smtp.setSubject(msg_subject);
	
	                        buffer = input();
	                        while( !buffer.equals(".") )
	                        {
	                            msg += buffer;
	                            msg += "\n";
	                            buffer = input();
	                        }
	
	                        msg += msg_from;  // insert name of previous message author
	                        msg += " wrote: \n";
	                        msg += reply;     // insert previous message
	                        reply = "";      // reset reply field for future use
	
	                        smtp.setMessageText(msg);
	                        System.out.print( "Cc: " );
	                        buffer = input();
	                        smtp.setCc(buffer);
	                        smtp.send();
	                    }
	                }
	                else if ( argument[0].equals("v") )
	                {
	                    if ( argument.length > 1 )
	                    {
	                        msgnum = Integer.parseInt(argument[1]);
	                        state = FULL;
	                        pop.setMessageNumber(msgnum);
	                        pop.retrieve();
	                    }
	                    else
	                    {
	                        System.out.println( "No message specified." );
	                    }
	                }
	                else
	                {
	                    try
	                    {
	                        // allow user to enter only the number
	                        //  of the message they want to view
	                        msgnum = Integer.parseInt(command);
	                        state = FULL;
	                        pop.setMessageNumber(msgnum);
	                        pop.retrieve();
	                    }
	                    catch(NumberFormatException e)
	                    {
	                        System.out.println( "Bad command / Not implemented in demo." );
	                    }
	                } // end of command checking
	            }  // end of main while loop
	        }
	        catch(Exception e)
	        {
	        	displayError(e);
	        }
		}        
    }

    private static void DisplayMenu() {
    	System.out.println("Mail Commands");
        System.out.println("v <message number>     view the content of selected message");
        System.out.println("n                      goto and view next message");
        System.out.println("h                      give head lines of messages");
        System.out.println("d <message number>     delete selected message");
        System.out.println("m <user>               mail to specific users");
        System.out.println("q                      quit, saving unresolved messages in mbox");       
        System.out.println("r <message number>     reply to message");	
	}   
}

class ConsoleDemo {
  private static BufferedReader bf = new BufferedReader(new InputStreamReader(System.in));

  static String input() {
    try {
      return bf.readLine();
    } catch (IOException ioe) {
      return "";
    }
  }
  static char read() {
    return input().charAt(0);
  }

  static String prompt(String label) {
    return prompt(label, ":");
  }
  static String prompt(String label, String punctuation) {
    System.out.print(label + punctuation + " ");
    return input();
  }
  static String prompt(String label, String punctuation, String defaultVal) {
      System.out.print(label + " [" + defaultVal + "]" + punctuation + " ");
      String response = input();
      if (response.equals(""))
        return defaultVal;
      else
        return response;
  }

  static char ask(String label) {
    return ask(label, "?");
  }
  static char ask(String label, String punctuation) {
    return ask(label, punctuation, "(y/n)");
  }
  static char ask(String label, String punctuation, String answers) {
    System.out.print(label + punctuation + " " + answers + " ");
    return Character.toLowerCase(read());
  }

  static void displayError(Exception e) {
    System.out.print("Error");
    if (e instanceof IPWorksException) {
      System.out.print(" (" + ((IPWorksException) e).getCode() + ")");
    }
    System.out.println(": " + e.getMessage());
    e.printStackTrace();
  }

  /**
   * Takes a list of switch arguments or name-value arguments and turns it into a map.
   */
  static java.util.Map<String, String> parseArgs(String[] args) {
    java.util.Map<String, String> map = new java.util.HashMap<String, String>();
    
    for (int i = 0; i < args.length; i++) {
      // Add a key to the map for each argument.
      if (args[i].startsWith("-")) {
        // If the next argument does NOT start with a "-" then it is a value.
        if (i + 1 < args.length && !args[i + 1].startsWith("-")) {
          // Save the value and skip the next entry in the list of arguments.
          map.put(args[i].toLowerCase().replaceFirst("^-+", ""), args[i + 1]);
          i++;
        } else {
          // If the next argument starts with a "-", then we assume the current one is a switch.
          map.put(args[i].toLowerCase().replaceFirst("^-+", ""), "");
        }
      } else {
        // If the argument does not start with a "-", store the argument based on the index.
        map.put(Integer.toString(i), args[i].toLowerCase());
      }
    }
    return map;
  }
}



